### ISSUE TITLE
Use of nanosleep (in dds_sleepfor) is unwise

### ISSUE THREAD

**Issue by rburchell** [eclipse-cyclonedds/cyclonedds#425]
Not a very important issue, but while browsing code I noticed that you're using nanosleep to sleep for a given duration. This function is widely available (so for portability, and a fallback) it's definitely a good choice, but it is potentially unfortunate from a reliability perspective.

nanosleep() uses a relative time interval, and that relative time is measured against the realtime clock on some platforms, which could mean that time skew can theoretically be a problem: I've seen non-mainstream platforms where this has been a problem in the past.

You might want to consider looking at clock_nanosleep (if available) with CLOCK_MONOTONIC to be explicit about the requirements for this, perhaps?

**Comment by eboasson** [comment]
@rburchell you raise an good point. I think the nanosleep part is a relatively small problem because outside the tests (and ddsperf/pubsub/helloworld/...) it is only used in places where it wants "a short sleep", and some of those are even stop-gap measures. Still, it would be an improvement to use clock_nanosleep.

Looking a bit further than nanosleep, there's pthread_cond_timedwait, on which nearly all the timed things are built. For example, the timed event handling is built on the monotonic clock but it currently needs to convert a wake-up time in that clock domain to a wall-clock times. Consistently using the correct clock via the interfaces (at least) Linux provides would be a major improvement.

Without your raising this now, it might have been a while before anyone noticed. Thanks!

**Comment by rburchell** [comment]
Yes, absolutely, I don't see this as being high priority. But when it does go wrong (and with time, that seems to be an inevitable "when", rather than an "if"), it can be quite annoying, and even more so to track down, so I figured I'd push an issue just so it was considered. :-)
